File: backend/internal/model/requirement.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Requirement 需求表
type Requirement struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Title       string `gorm:"size:200;not null" json:"title"`        // 需求标题
	Description string `gorm:"type:text" json:"description"`        // 需求描述（Markdown）
	Status      string `gorm:"size:20;default:'draft'" json:"status"` // 状态：draft(草稿), reviewing(评审中), active(激活), changing(变更中), closed(已关闭)
	Priority    string `gorm:"size:20;default:'medium'" json:"priority"` // 优先级：low, medium, high, urgent

	ProjectID uint    `gorm:"index;not null" json:"project_id"` // 必填关联项目
	Project   Project `gorm:"foreignKey:ProjectID" json:"project,omitempty"`

	CreatorID uint `gorm:"index" json:"creator_id"`
	Creator   User `gorm:"foreignKey:CreatorID" json:"creator,omitempty"`

	AssigneeID *uint `gorm:"index" json:"assignee_id"`
	Assignee   *User `gorm:"foreignKey:AssigneeID" json:"assignee,omitempty"`

	EstimatedHours *float64 `gorm:"default:0" json:"estimated_hours"` // 预估工时（小时）
	ActualHours    *float64 `gorm:"default:0" json:"actual_hours"`    // 实际工时（小时），从资源分配自动计算
}

// Bug Bug表
type Bug struct {
	ID        uint           `gorm:"primarykey" json:"id"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Title       string `gorm:"size:200;not null" json:"title"`        // Bug标题
	Description string `gorm:"type:text" json:"description"`         // Bug描述（Markdown）
	Status      string `gorm:"size:20;default:'active'" json:"status"` // 状态：active(激活), resolved(已解决), closed(已关闭)
	Priority    string `gorm:"size:20;default:'medium'" json:"priority"` // 优先级：low, medium, high, urgent
	Severity    string `gorm:"size:20;default:'medium'" json:"severity"` // 严重程度：low, medium, high, critical

	ProjectID uint    `gorm:"index;not null" json:"project_id"`
	Project   Project `gorm:"foreignKey:ProjectID" json:"project,omitempty"`

	CreatorID uint `gorm:"index" json:"creator_id"`
	Creator   User `gorm:"foreignKey:CreatorID" json:"creator,omitempty"`

	Assignees []User `gorm:"many2many:bug_assignees;" json:"assignees,omitempty"`

	RequirementID *uint       `gorm:"index" json:"requirement_id"` // 关联需求
	Requirement   *Requirement `gorm:"foreignKey:RequirementID" json:"requirement,omitempty"`

	ModuleID *uint  `gorm:"index" json:"module_id"` // 关联功能模块
	Module   *Module `gorm:"foreignKey:ModuleID" json:"module,omitempty"`

	EstimatedHours *float64 `gorm:"default:0" json:"estimated_hours"` // 预估工时（小时）
	ActualHours    *float64 `gorm:"default:0" json:"actual_hours"`    // 实际工时（小时），从资源分配自动计算

	// 解决方案相关
	Solution       string `gorm:"size:50" json:"solution"`        // 解决方案：设计如此、重复Bug、外部原因、已解决、无法重现、延期处理、不予解决、转为研发需求
	SolutionNote   string `gorm:"type:text" json:"solution_note"` // 解决方案备注
	ResolvedVersionID *uint   `gorm:"index" json:"resolved_version_id"` // 解决版本ID
	ResolvedVersion   *Version `gorm:"foreignKey:ResolvedVersionID" json:"resolved_version,omitempty"`
}

// BugAssignee Bug分配表
type BugAssignee struct {
	BugID    uint `gorm:"primaryKey" json:"bug_id"`
	UserID   uint `gorm:"primaryKey" json:"user_id"`
	AssignedAt time.Time `gorm:"default:CURRENT_TIMESTAMP" json:"assigned_at"`
}

File: frontend/src/api/requirement.ts
import request from '../utils/request'

export interface Requirement {
  id: number
  title: string
  description?: string
  status: 'draft' | 'reviewing' | 'active' | 'changing' | 'closed'
  priority: 'low' | 'medium' | 'high' | 'urgent'
  project_id: number // 必填
  project?: any
  creator_id: number
  creator?: any
  assignee_id?: number
  assignee?: any
  estimated_hours?: number
  actual_hours?: number
  created_at?: string
  updated_at?: string
}

export interface RequirementListResponse {
  list: Requirement[]
  total: number
  page: number
  page_size: number
}

export interface CreateRequirementRequest {
  title: string
  description?: string
  status?: 'draft' | 'reviewing' | 'active' | 'changing' | 'closed'
  priority?: 'low' | 'medium' | 'high' | 'urgent'
  project_id: number // 必填
  assignee_id?: number
  estimated_hours?: number
}

export interface UpdateRequirementStatusRequest {
  status: 'draft' | 'reviewing' | 'active' | 'changing' | 'closed'
}

// 需求相关API
export const getRequirements = async (params?: {
  keyword?: string
  project_id?: number
  status?: string
  priority?: string
  assignee_id?: number
  creator_id?: number
  page?: number
  size?: number
}): Promise<RequirementListResponse> => {
  return request.get('/requirements', { params })
}

export const getRequirement = async (id: number): Promise<Requirement> => {
  return request.get(`/requirements/${id}`)
}

export const createRequirement = async (data: CreateRequirementRequest): Promise<Requirement> => {
  return request.post('/requirements', data)
}

export const updateRequirement = async (id: number, data: Partial<CreateRequirementRequest>): Promise<Requirement> => {
  return request.put(`/requirements/${id}`, data)
}

export const deleteRequirement = async (id: number): Promise<void> => {
  return request.delete(`/requirements/${id}`)
}

export const updateRequirementStatus = async (id: number, data: UpdateRequirementStatusRequest): Promise<Requirement> => {
  return request.patch(`/requirements/${id}/status`, data)
}

export interface RequirementStatistics {
  total: number
  pending: number
  in_progress: number
  completed: number
  cancelled: number
  low_priority: number
  medium_priority: number
  high_priority: number
  urgent_priority: number
}

export const getRequirementStatistics = async (params?: {
  keyword?: string
  project_id?: number
  assignee_id?: number
  creator_id?: number
}): Promise<RequirementStatistics> => {
  return request.get('/requirements/statistics', { params })
}

// 历史记录相关接口（参考bug.ts和project.ts）
export interface History {
  id: number
  action_id: number
  field: string
  old: string
  old_value: string
  new: string
  new_value: string
  diff?: string
  created_at?: string
}

export interface Action {
  id: number
  object_type: string
  object_id: number
  project_id: number
  actor_id: number
  actor?: {
    id: number
    username: string
    nickname?: string
  }
  action: 'created' | 'edited' | 'assigned' | 'resolved' | 'closed' | 'confirmed' | 'commented' | 'status_changed'
  date: string
  comment?: string
  extra?: string
  histories?: History[]
  created_at?: string
}

export interface RequirementHistoryResponse {
  list: Action[]
}

export interface AddRequirementHistoryNoteRequest {
  comment: string
}

export const getRequirementHistory = async (id: number): Promise<RequirementHistoryResponse> => {
  return request.get(`/requirements/${id}/history`)
}

export const addRequirementHistoryNote = async (id: number, data: AddRequirementHistoryNoteRequest): Promise<{ message: string }> => {
  return request.post(`/requirements/${id}/history/note`, data)
}

